megamek/src/megamek/client/bot/princess/Princess.java

```diff
@@ -629,10 +631,152 @@ private int calculateTurretDeploymentValue(final Coords coords) {
+        // Logging is extremely slow, only use when debugging.
+        StringBuilder sb = null;
+        if (logger.isDebugEnabled()) {
+            sb = new StringBuilder();
+            sb.append("Ranking kernel around hex ").append(start.getFinalCoords().toString());
+        }
+
+        double rank;
+        // allAtDistance uses a concept of radius that is 1 smaller.
+        ArrayList<Coords> kernel = start.getFinalCoords().allAtDistance(radius+1);
+
+        // Get all paths from the start point to the outer hexes that use "radius" MP
+        // Worse starting hexes have fewer, and shorter, paths
+        ShortestPathFinder pf = ShortestPathFinder.newInstanceOfOneToAll(
+                                        radius, MoveStepType.FORWARDS, game);
+        pf.run(start);
+
+        // Lower rank is better; 0.0 is minimum at this point.
+        rank = Math.max(kernel.size() - pf.getAllComputedPaths().size(), 0.0);
+        for (MovePath mp: pf.getAllComputedPaths().values()) {
+            rank -= mp.getHexesMoved();
+            rank += ranker.checkPathForHazards(mp, deployedUnit, game);
+        }
+
+        if (sb != null) {
+            sb.append("\n\tAll computed ")
+                    .append(radius)
+                    .append("-length paths: ")
+                    .append(pf.getAllComputedPaths().size());
+            sb.append("\n\tFinal rank (including hazards): ").append(rank);
+            logger.debug(sb.toString());
+        }
+
+        return rank;
+    }
+
+    /**
+     * Rank possible deployment coordinates by hazard, path freedom, concealment
+     * 1. Randomly select N coords from list
+     * 2. For selected coords:
+     *      1. Check if hex is invalid
+     *      2. Create a MovePath containing the starting coordinate
+     *      3. Get the hazard value
+     *      4. Save Coords to HashMap with hazard as key
+     *
+     * @param deployedUnit
+     * @param possibleDeployCoords
+     * @return
+     */
+    protected Coords rankDeploymentCoords(Entity deployedUnit, List<Coords> possibleDeployCoords) {
+        StringBuilder sb = null;
+        if (logger.isDebugEnabled()) {
+            sb = new StringBuilder();
+            sb.append("Ranking deployment hexes...");
+        }
+
+        // Sample LIMIT number of valid starting hexes, check accessibility and hazards within RADIUS
+        int LIMIT = 20;
+        int RADIUS = 3;
+
+        // Shallow copy of refs list
+        ArrayList<Coords> localCopy = new ArrayList(possibleDeployCoords);
+        // Shuffle
+        Collections.shuffle(localCopy);
+
+        // Hacky, but really, "DEPLOY" should be a path step...
+        MovePath mp = new MovePath(game, deployedUnit);
+        mp.addStep(MoveStepType.NONE);
+        MoveStep deployStep = mp.getLastStep();
+        IPathRanker ranker = getPathRanker(deployedUnit);
+        HashMap<Double, ArrayList<Coords>> rankedCoords = new HashMap<>();
+
+        if (!deployedUnit.isAero()) {
```

Comment:
Probably add VTOL and WIGE to this as well.

Reply:
Do we deploy WiGEs in flight?  If not, they probably still want all the hazard checks...

Reply:
It's optional to deploy WiGEs in flight, so WiGEs that _don't_ will want the extra checks.

Reply:
We do weight cul-de-sac hexes lower, and completely inaccessible hexes lower still, but it's fairly unsubtle.

After reviewing the botClient code, though, I am thinking we can remove the shuffle before sampling - the list of valid coordinates should be in a semblance of fitness order, so just iterating through the first 20 (+/- some) should be fine.

I'll test this.

Reply:
Looks like _not_ re-shuffling leads to slightly less clumping, so let's go with that.